HPOS - Update WooCommerceSubscriptions segment filter [MAILPOET-4568]

costasovo opened the pull request:
## Description

This PR adds support for HPOS to the WooCommerce Subscriptions dynamic segment filter.

## Code review notes
I tried to explain the changes in commit messages.

## QA notes
1) Install Woo and WooCommerce Subscriptions.
2) Create a subscription product (Hint: you can set price 0 and then subscribe without configuring payments)
3) Subscribe to the subscription product and also confirm the subscription to MailPoet emails (you need a subscribed subscriber)
4) Create a dynamic segment for the subscription product
5) You should see one subscriber
6) Enable HPOS on Woo (Settings > Advanced > Features) and wait for orders to migrate (threre is info about migration status in settings).
7) Check that the dynamic segment still contains the subscriber.

You can also switch to HPOS immediately after installing Woo. So you don't need to wait for migration.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4568]

## After-merge notes

_N/A_


[MAILPOET-4568]: https://mailpoet.atlassian.net/browse/MAILPOET-4568?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Refactor WooSubscriptions integration test to use factory instead post

The factory uses wcs_create_subscription and creates the subscription order
in proper table according to HPOS setting.
[MAILPOET-4568]

costasovo committed:
Enable WooSubscriptions test when HPOS is active

We need to temporarily turn off plugin compatibility check in Woo
because current Woo Subscriptions are still officially not marked as
HPOS compatible.
[MAILPOET-4568]

costasovo committed:
Update WooSubscription filter to work with HPOS data

The wc_orders table doesn't contain user_id but has billing_email and customer_id.
These were two options I could use for joining subscribers table.
I could use the customer_id but I would have to join wc_customer_lookup table and then
the orders table. Instead of customer_id I chose the billing_email. Both tables have
indexes on the email columns and it is one join instead of two.
[MAILPOET-4568]

costasovo committed:
Remove temporary action for disabling Woo compatibility check in test

WooSubscriptions declared compatibility with HPOS in version 4.9.0
[MAILPOET-4568]

costasovo committed:
Update WooCommerce and plugins in tests on CI

We need to use HPOS compatible Woo Subscriptions
[MAILPOET-4568]

costasovo committed:
Re-enable acceptance test for Woo Subscriptions segment

[MAILPOET-4568]

costasovo committed:
Add collation check to WooSubscriptions filter

We join using a string column and it may cause errors when the columns
using different collation. This commit adds collation check so that we can alter
collation in the join condition in case it is needed.
[MAILPOET-4568]

triple0t commented on mailpoet/tests/DataFactories/WooCommerceSubscription.php:
```diff
@@ -3,7 +3,7 @@
 
 class WooCommerceSubscription {
-  public function createSubscription(int $userId, int $subscriptionProductId): \WC_Subscription {
+  public function createSubscription(int $userId, int $subscriptionProductId, string $status = 'active'): \WC_Subscription {
     $args = [
       'status' => 'active',
```
Do you plan to use the `$status` parameter in the `$args` array?

costasovo replied:
Great catch! Thank you!

triple0t commented on mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php:
```diff
@@ -33,26 +33,18 @@ class WooCommerceSubscriptionTest extends \MailPoetTest {
-      );
+
+      $status = explode('_', $email)[0];
+      $this->subscriptionsFactory ->createSubscription($userId, $productId, $status);
```
There's a space here between `subscriptionsFactory` and `->createSubscription`

triple0t reviewed (commented):
Looks good.
I have just two comments. Kindly locate them below.

costasovo committed:
Fix usage of status parameter in Woo Subscriptions data factory

[MAILPOET-4568]

costasovo committed:
Fix coding style issue in Woo Subscriptions filter test

[MAILPOET-4568]

costasovo commented:
Thank you @triple0t for the review and for caching the issues! I added two new commits addressing the issues. Could you please take another look?

triple0t reviewed (dismissed):
LGTM

Aschepikov commented:
Hi @costasovo there are conflicts, can you resolve them pls. Also, I want to add that I can't test it with hpos because there is no compatibility with woo subscriptions and there is no subscribers in the dynamic segment if the hpos is enabled

costasovo commented:
@Aschepikov I fixed the conflict and rebased. If I'm not mistaken,  Woo Subscription declares compatibility with HPOS from version 4.9.1 +. If you don't have access to the plugin I can send you the zip. Just ping me in DM.

Aschepikov reviewed (approved):
GJ!
